File: MosAlgorithm.java
import java.util.*;
import java.io.*;

public class MosAlgorithm {
    static class Query implements Comparable<Query> {
        int l, r, idx;

        Query(int l, int r, int idx) {
            this.l = l;
            this.r = r;
            this.idx = idx;
        }

        @Override
        public int compareTo(Query other) {
            int blockA = this.l / blockSize;
            int blockB = other.l / blockSize;
            if (blockA != blockB)
                return Integer.compare(blockA, blockB);
            return Integer.compare(this.r, other.r);
        }
    }

    static int[] arr;
    static int[] freq;
    static int[] answers;
    static int distinct = 0;
    static int blockSize;

    static void add(int index) {
        freq[arr[index]]++;
        if (freq[arr[index]] == 1)
            distinct++;
    }

    static void remove(int index) {
        freq[arr[index]]--;
        if (freq[arr[index]] == 0)
            distinct--;
    }

    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt(); // size of array
        int q = sc.nextInt(); // number of queries

        arr = new int[n];
        freq = new int[1000001]; // assuming elements <= 1e6
        blockSize = (int) Math.sqrt(n);

        for (int i = 0; i < n; i++)
            arr[i] = sc.nextInt();

        Query[] queries = new Query[q];
        answers = new int[q];

        for (int i = 0; i < q; i++) {
            int l = sc.nextInt();
            int r = sc.nextInt();
            queries[i] = new Query(l, r, i);
        }

        Arrays.sort(queries);

        int l = 0, r = -1;

        for (Query query : queries) {
            while (l > query.l) add(--l);
            while (r < query.r) add(++r);
            while (l < query.l) remove(l++);
            while (r > query.r) remove(r--);

            answers[query.idx] = distinct;
        }

        for (int ans : answers)
            System.out.println(ans);
    }
}

File: MosAlgorithm.cpp
#include <bits/stdc++.h>
using namespace std;

const int N = 1e5 + 10;
int arr[N];
int freq[N];        // Frequency of each element
int answers[N];     // Answers to the queries
int distinctCount = 0;

struct Query {
    int l, r, idx;
};

int blockSize;

bool compare(Query a, Query b) {
    int blockA = a.l / blockSize;
    int blockB = b.l / blockSize;
    if (blockA != blockB)
        return blockA < blockB;
    return a.r < b.r;  // Optional optimization: blockA % 2 ? a.r > b.r : a.r < b.r;
}

void add(int index) {
    freq[arr[index]]++;
    if (freq[arr[index]] == 1)
        distinctCount++;
}

void remove(int index) {
    freq[arr[index]]--;
    if (freq[arr[index]] == 0)
        distinctCount--;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n, q;
    cin >> n >> q;

    blockSize = sqrt(n);

    for (int i = 0; i < n; ++i)
        cin >> arr[i];

    vector<Query> queries(q);
    for (int i = 0; i < q; ++i) {
        cin >> queries[i].l >> queries[i].r;
        queries[i].idx = i;
    }

    sort(queries.begin(), queries.end(), compare);

    int currL = 0, currR = -1;

    for (Query query : queries) {
        int L = query.l;
        int R = query.r;

        while (currL > L) add(--currL);
        while (currR < R) add(++currR);
        while (currL < L) remove(currL++);
        while (currR > R) remove(currR--);

        answers[query.idx] = distinctCount;
    }

    for (int i = 0; i < q; ++i)
        cout << answers[i] << "\n";

    return 0;
}
